### ISSUE TITLE
index.json in S3 grows indefinitely and causes errors

### ISSUE THREAD

**Issue by dylanmoz** [craftship/codebox-npm#84]
## Description

Publishing a package that is 3Mb will be saved to s3 on its own but also to the index.json:

(src/put/publish.js)
```js
    json['dist-tags'][tag] = version;
    json._attachments[`${name}-${version}.tgz`] = pkg._attachments[`${name}-${version}.tgz`];
    json.versions[version] = versionData;

    ...

    await storage.put(
      `${name}/${version}.tgz`,
      json._attachments[`${name}-${version}.tgz`].data, // eslint-disable-line no-underscore-dangle
      'base64',
    );

    await storage.put(
      `${name}/index.json`,
      JSON.stringify(json),
    );
```

If you publish 100 times, the index.json will be roughly 300Mb, which will fail or be grossly inefficient:
```
    const pkgBuffer = await storage.get(`${name}/index.json`);
    json = JSON.parse(pkgBuffer.toString());
```

Would the solution be to clear out `json._attachements` prior to saving to index.json?
What recommendations do you have since I have already run into the issue, can I simply delete the entirety of the bucket in s3 if I don't care about past releases?

**Comment by jonsharratt** [comment]
Sorry for the delay getting back to you, which call is causing you errors, there is a hard limit we are getting problems with #39 when the json without attachments is over 6MB.  The GET request removes attachments when serving it for this reason.

https://github.com/craftship/codebox-npm/blob/master/src/get/lib.js#L13

As you mention it might be time to remove this, this was an original npm way of working as far as I understand and they no doubt realised this is not a maintainable solution going forward.

Need to understand it a little further as would not expect it to cause you any issues apart from large storage sizes.

**Comment by dylanmoz** [comment]
@jonsharratt I get a 502 error when running `npm publish`, so on the PUT to the registry I believe.

**Comment by k-k** [comment]
@jonsharratt Im guessing the get request removed attachments, but does the PUT request attempt to return the patched index.json as a response - and does not remove attachments?

**Comment by ganapativs** [comment]
@jonsharratt @kmfk Yup, even in my package, index.json was like 58mb, so, the publishing of new package version was throwing 502 & then 403(Probably Lambda couldn't handle it?). 

I had to cleanup previous attachments while publishing to fix the issue(in publish.js). 
This reduced index.json size to 1.7mb and publishing and install are working fine.

Pull request here - https://github.com/craftship/codebox-npm/issues/84.

Just curious, why we are storing all attachment data in index.json?

**Comment by jonsharratt** [comment]
So this existed as I had to piece together how npm works.  Turns out they used couchdb so this was why I copied over that functionallity.

They hit the same perf issues and look to have taken the attachments out of the package.json themselves.

Merged your proposed fix @ganapativs and tagged a release, give it a go and let me know how you get on.

Sorry for the delay on this been a crazy past couple of months.
